crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java

```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.crypto.scrypt;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.UUID;
+
+import org.junit.Test;
+
+/**
+ * @author Shazin Sadakath
+ *
+ */
+public class SCryptPasswordEncoderTests {
+
+    @Test
+    public void matches() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        String result = encoder.encode("password");
+        assertFalse(result.equals("password"));
+        assertTrue(encoder.matches("password", result));
+    }
+
+    @Test
+    public void unicode() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        String result = encoder.encode("passw\u9292rd");
+        assertFalse(encoder.matches("pass\u9292\u9292rd", result));
+        assertTrue(encoder.matches("passw\u9292rd", result));
+    }
+
+    @Test
+    public void notMatches() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        String result = encoder.encode("password");
+        assertFalse(encoder.matches("bogus", result));
+    }
+    
+    @Test
+    public void customParameters() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder(UUID.randomUUID().toString(), 512, 8, 4);
+        String result = encoder.encode("password");
+        assertFalse(result.equals("password"));
+        assertTrue(encoder.matches("password", result));
+    }
+
+    @Test
+    public void doesntMatchNullEncodedValue() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        assertFalse(encoder.matches("password", null));
+    }
+
+    @Test
+    public void doesntMatchEmptyEncodedValue() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        assertFalse(encoder.matches("password", ""));
+    }
+
+    @Test
+    public void doesntMatchBogusEncodedValue() {
+        SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
+        assertFalse(encoder.matches("password", "012345678901234567890123456789"));
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidSaltParameter() {
+        new SCryptPasswordEncoder(null, Integer.MIN_VALUE, 16, 2);     
+    } 
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidCpuCostParameter() {
+        new SCryptPasswordEncoder(UUID.randomUUID().toString(), Integer.MIN_VALUE, 16, 2);     
+    }   
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidMemoryCostParameter() {
+        new SCryptPasswordEncoder(UUID.randomUUID().toString(), 2, Integer.MAX_VALUE, 2);     
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidParallelizationParameter() {
+        new SCryptPasswordEncoder(UUID.randomUUID().toString(), 2, 8, Integer.MAX_VALUE);     
+    }
+

```

Comment:
Add a test that encoding the same password twice results in different results (due to a random salt being used).

```java
SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
String password = "secret";
assertThat(encoder.encode(password)).isNotEqualTo(encoder.encode(secret));
```